plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py

```diff
@@ -334,6 +334,25 @@ def __init__(self, task_config: Elastic, task_function: Callable, **kwargs):
                 self.pod_template = PodTemplate()
             add_shared_mem_volume_to_pod_template(self.pod_template)
 
+    def get_error_file_name_suffix(self) -> Optional[str]:
+        """
+        Returns error file name suffix that represents the replica index.
+        """
+        # This is fragile, as it assumes hostname == podname, which is not the case if
```

Comment:
What's needed is to customize the pod spec to set an environment variable that captures the value of metadata.labels."training.kubeflow.org/replica-index" :

```
env:
 - name: FLYTE_PYTORCH_TASK_REPLICA_INDEX
   valueFrom:
     fieldRef:
       fieldPath: metadata.labels['training.kubeflow.org/replica-index']
```

Reply:
Setting this env var via the downward api in the pod spec in flyteplugins would in theory work I'd say 👍
I agree that we shouldn't use the hostname.

I have two questions though:

* Does the error file name actually need to contain the group rank? If the error message itself would contain the info which group rank failed, could we get away with a random uuid suffix which just prevents overwriting?
* Do you know how torch distributed elastic determines the group rank for the worker pods/processes? I just checked on a running elastic task and (in line with what you said) only see the `PET_NNODES` and `PET_NPROC_PER_NODE` env vars which specify the number of nodes and processes but not the rank. Could it be that `worker-0` ends up being group rank 2 or how is this decided at rendezvous?

Reply:
The ranks are decided by the rendezvous and worker-0 is not necessarily the one that has rank 0 or group rank 0.

I agree that the suffix does not need to be the group rank. We can do uuid, but internal debugging might get a bit harder, as by looking at the error files, no one would know what pod actually generated it.

What about using the pod name? Is that available in Flyte via some mechanism or does that also require downward api usage?

Reply:
> The ranks are decided by the rendezvous and worker-0 is not necessarily the one that has rank 0 or group rank 0.

Ok then it makes sense 👍 

> What about using the pod name? Is that available in Flyte via some mechanism or does that also require downward api usage?

I think it's not available unfortunately. Could be added though I'd say.

> I agree that the suffix does not need to be the group rank. We can do uuid, but internal debugging might get a bit harder, as by looking at the error files, no one would know what pod actually generated it

The [`ChiledFailedError`](https://github.com/pytorch/pytorch/blob/f3df7deab8953af76ff1723ed49094208057a834/torch/distributed/elastic/multiprocessing/errors/__init__.py#L234) has a ref to the global rank but that doesn't help us because it's not linked to the pod and we are interested in the pod name in order to look at the respective stackdriver logs for more info - correct?

Reply:
`ChiledFailedError` is not the only error that comes out of `torch_elastic`, unfortunately. 

We can start with uuid and improve afterwards. Made the change.